File: src/main/java/com/engine/nnue_trainer/v2/DenseFeatures.java
package com.engine.nnue_trainer.v2;

import com.engine.nnue_trainer.board.Board;
import com.engine.nnue_trainer.board.Cell;
import com.engine.nnue_trainer.board.CellKind;
import com.engine.nnue_trainer.board.Pos;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Set;

public class DenseFeatures {

  public static float[] extract(Board board, int activePlayer, int turnNumber) {
    int opponent = 3 - activePlayer;

    int stmNormal = 0;
    int nstmNormal = 0;
    int stmFortified = 0;
    int nstmFortified = 0;
    int neutral = 0;
    int empty = 0;

    float stmBaseAlive = 0.0f;
    float nstmBaseAlive = 0.0f;

    List<Pos> stmCells = new ArrayList<>();
    List<Pos> nstmCells = new ArrayList<>();

    for (int r = 0; r < board.rows; r++) {
      for (int c = 0; c < board.cols; c++) {
        Cell cell = board.getCell(r, c);
        if (cell == null || cell.kind == CellKind.EMPTY) {
          empty++;
        } else if (cell.kind == CellKind.NEUTRAL) {
          neutral++;
        } else if (cell.kind == CellKind.NORMAL) {
          if (cell.owner == activePlayer) {
            stmNormal++;
            stmCells.add(new Pos(r, c));
          } else if (cell.owner == opponent) {
            nstmNormal++;
            nstmCells.add(new Pos(r, c));
          }
        } else if (cell.kind == CellKind.FORTIFIED) {
          if (cell.owner == activePlayer) {
            stmFortified++;
            stmCells.add(new Pos(r, c));
          } else if (cell.owner == opponent) {
            nstmFortified++;
            nstmCells.add(new Pos(r, c));
          }
        } else if (cell.kind == CellKind.BASE) {
          if (cell.owner == activePlayer) {
            stmBaseAlive = 1.0f;
            stmCells.add(new Pos(r, c));
          } else if (cell.owner == opponent) {
            nstmBaseAlive = 1.0f;
            nstmCells.add(new Pos(r, c));
          }
        }
      }
    }

    float maxDist = board.rows + board.cols - 2.0f;
    float totalArea = board.rows * board.cols;

    float stmMinDist = maxDist;
    float nstmMinDist = maxDist;

    Pos nstmBasePos = getBasePos(opponent, board.rows, board.cols);
    if (nstmBasePos != null) {
      for (Pos p : stmCells) {
        float d = manhattanDist(p.row, p.col, nstmBasePos.row, nstmBasePos.col);
        if (d < stmMinDist) {
          stmMinDist = d;
        }
      }
    }

    Pos stmBasePos = getBasePos(activePlayer, board.rows, board.cols);
    if (stmBasePos != null) {
      for (Pos p : nstmCells) {
        float d = manhattanDist(p.row, p.col, stmBasePos.row, stmBasePos.col);
        if (d < nstmMinDist) {
          nstmMinDist = d;
        }
      }
    }

    float stmComponents = getComponents(stmCells);
    float nstmComponents = getComponents(nstmCells);

    float stmCompRatio = stmCells.isEmpty() ? 0.0f : stmComponents / stmCells.size();
    float nstmCompRatio = nstmCells.isEmpty() ? 0.0f : nstmComponents / nstmCells.size();

    float[] features = new float[14];
    features[0] = stmNormal / totalArea;
    features[1] = nstmNormal / totalArea;
    features[2] = stmFortified / totalArea;
    features[3] = nstmFortified / totalArea;
    features[4] = neutral / totalArea;
    features[5] = empty / totalArea;
    features[6] = stmBaseAlive;
    features[7] = nstmBaseAlive;
    features[8] = stmMinDist / maxDist;
    features[9] = nstmMinDist / maxDist;
    features[10] = stmCompRatio;
    features[11] = nstmCompRatio;
    features[12] = turnNumber / 100.0f;
    features[13] = totalArea / 144.0f;

    return features;
  }

  private static Pos getBasePos(int player, int rows, int cols) {
    if (player == 1) return new Pos(0, 0);
    if (player == 2) return new Pos(rows - 1, cols - 1);
    if (player == 3) return new Pos(0, cols - 1);
    if (player == 4) return new Pos(rows - 1, 0);
    return null;
  }

  private static int manhattanDist(int r1, int c1, int r2, int c2) {
    return Math.abs(r1 - r2) + Math.abs(c1 - c2);
  }

  private static int getComponents(List<Pos> cells) {
    if (cells.isEmpty()) return 0;

    Set<Pos> visited = new HashSet<>();
    Set<Pos> cellSet = new HashSet<>(cells);
    int components = 0;

    int[] dr = {-1, -1, -1, 0, 0, 1, 1, 1};
    int[] dc = {-1, 0, 1, -1, 1, -1, 0, 1};

    for (Pos startCell : cells) {
      if (!visited.contains(startCell)) {
        components++;
        Queue<Pos> queue = new LinkedList<>();
        queue.add(startCell);
        visited.add(startCell);

        while (!queue.isEmpty()) {
          Pos curr = queue.poll();
          for (int i = 0; i < 8; i++) {
            Pos neighbor = new Pos(curr.row + dr[i], curr.col + dc[i]);
            if (cellSet.contains(neighbor) && !visited.contains(neighbor)) {
              visited.add(neighbor);
              queue.add(neighbor);
            }
          }
        }
      }
    }

    return components;
  }
}

File: python/v2/dense_features.py
def _get_base_pos(player, rows, cols):
    if player == 1:
        return (0, 0)
    elif player == 2:
        return (rows - 1, cols - 1)
    elif player == 3:
        return (0, cols - 1)
    elif player == 4:
        return (rows - 1, 0)
    return None

def _manhattan_dist(r1, c1, r2, c2):
    return abs(r1 - r2) + abs(c1 - c2)

def extract_dense_features(board, active_player, turn_number, rows=12, cols=12):
    opponent = 3 - active_player

    stm_normal = 0
    nstm_normal = 0
    stm_fortified = 0
    nstm_fortified = 0
    neutral = 0
    empty = 0

    stm_base_alive = 0.0
    nstm_base_alive = 0.0

    stm_cells = []
    nstm_cells = []

    for r in range(rows):
        for c in range(cols):
            cell = board[r][c]
            if cell is None or cell['kind'] == 'EMPTY':
                empty += 1
            elif cell['kind'] == 'NEUTRAL':
                neutral += 1
            elif cell['kind'] == 'NORMAL':
                if cell['owner'] == active_player:
                    stm_normal += 1
                    stm_cells.append((r, c))
                elif cell['owner'] == opponent:
                    nstm_normal += 1
                    nstm_cells.append((r, c))
            elif cell['kind'] == 'FORTIFIED':
                if cell['owner'] == active_player:
                    stm_fortified += 1
                    stm_cells.append((r, c))
                elif cell['owner'] == opponent:
                    nstm_fortified += 1
                    nstm_cells.append((r, c))
            elif cell['kind'] == 'BASE':
                if cell['owner'] == active_player:
                    stm_base_alive = 1.0
                    stm_cells.append((r, c))
                elif cell['owner'] == opponent:
                    nstm_base_alive = 1.0
                    nstm_cells.append((r, c))

    max_dist = float(rows + cols - 2)
    total_area = float(rows * cols)

    # distance to enemy base
    stm_min_dist = max_dist
    nstm_min_dist = max_dist

    nstm_base_pos = _get_base_pos(opponent, rows, cols)
    if nstm_base_pos:
        for r, c in stm_cells:
            d = _manhattan_dist(r, c, nstm_base_pos[0], nstm_base_pos[1])
            if d < stm_min_dist:
                stm_min_dist = d

    stm_base_pos = _get_base_pos(active_player, rows, cols)
    if stm_base_pos:
        for r, c in nstm_cells:
            d = _manhattan_dist(r, c, stm_base_pos[0], stm_base_pos[1])
            if d < nstm_min_dist:
                nstm_min_dist = d

    # connected components
    def get_components(cells, owner):
        if not cells: return 0
        visited = set()
        components = 0
        cell_set = set(cells)
        directions = [(-1,-1), (-1,0), (-1,1), (0,-1), (0,1), (1,-1), (1,0), (1,1)]

        for start_cell in cells:
            if start_cell not in visited:
                components += 1
                queue = [start_cell]
                visited.add(start_cell)
                while queue:
                    cr, cc = queue.pop(0)
                    for dr, dc in directions:
                        nr, nc = cr + dr, cc + dc
                        if (nr, nc) in cell_set and (nr, nc) not in visited:
                            visited.add((nr, nc))
                            queue.append((nr, nc))
        return components

    stm_components = get_components(stm_cells, active_player)
    nstm_components = get_components(nstm_cells, opponent)

    stm_comp_ratio = stm_components / max(1.0, float(len(stm_cells)))
    nstm_comp_ratio = nstm_components / max(1.0, float(len(nstm_cells)))

    features = [
        float(stm_normal) / total_area,
        float(nstm_normal) / total_area,
        float(stm_fortified) / total_area,
        float(nstm_fortified) / total_area,
        float(neutral) / total_area,
        float(empty) / total_area,
        stm_base_alive,
        nstm_base_alive,
        float(stm_min_dist) / max_dist,
        float(nstm_min_dist) / max_dist,
        stm_comp_ratio,
        nstm_comp_ratio,
        float(turn_number) / 100.0,
        total_area / 144.0
    ]
    return features
